### PR TITLE
FIX: propagate dtype to idf_ for large corpuses in PandasNormalizedTfidfVectorizer

### PR DESCRIPTION
**Pull request by chrispyl**
### Issue
[Fitting](https://github.com/ing-bank/EntityMatchingModel/blob/8ce0939912319d85b4ea9fa436e93b5fc736eaef/emm/indexing/pandas_normalized_tfidf.py#L76) `PandasNormalizedTfidfVectorizer` for large _X_ returns results in `idf_` having `dtype` `np.float64` regardless of the provided `dtype`. This is due to `TfidfTransformer` which [assumes the resulting `dtype` of `np.log` and doesn't use the dtype parameter explicitly](https://github.com/scikit-learn/scikit-learn/blob/d5082d32de2797f9594c9477f2810c743560a1f1/sklearn/feature_extraction/text.py#L1666).
### Reproduction
```python
from sklearn.feature_extraction.text import TfidfVectorizer
import pandas as pd
import numpy as np
import uuid

#small data work fine
small_data=[str(uuid.uuid4()) for i in range(100)]
X = pd.Series(small_data)
vectorizer = TfidfVectorizer(dtype=np.float32)
vectorizer.fit(X)
print(vectorizer.idf_.dtype)

#large data do not preserve dtype
large_data=[str(uuid.uuid4()) for i in range(1000000)]
X = pd.Series(large_data)
vectorizer = TfidfVectorizer(dtype=np.float32)
vectorizer.fit(X)
print(vectorizer.idf_.dtype)
```
### Solution
Set the type of `idf_` right after fitting and before using it in any other operation to avoid surprises. This solution is temporary until the issue is fixed in scikit-learn.
### Sklearn issue background
Not propagating `dtype` to `TfidfTransformer` is a known [issue](https://github.com/scikit-learn/scikit-learn/pull/10443). Passing the `dtype` to `np.log` is not an option as in this thread they mention that it breaks functionality when `dtype` is integer. Adding: 
```pthon
.astype(dtype)
```
after the log operation was [proposed](https://github.com/scikit-learn/scikit-learn/pull/10443#discussion_r161215696) but it was removed.

Have opened a new [issue](https://github.com/scikit-learn/scikit-learn/issues/30016) in scikit-learn.

### CODE REVIEW

File: emm/indexing/pandas_normalized_tfidf.py [hunk 1]
```diff
@@ -37,8 +37,6 @@
 class PandasNormalizedTfidfVectorizer(TfidfVectorizer):
     """Implementation of customized TFIDF vectorizer"""
 
-    dtype = np.float32
```
Comment (mbaak) [inline emm/indexing/pandas_normalized_tfidf.py:40]: If we remove this, don't we need to set the default value (float32) elsewhere?

Comment (mbaak) [reply emm/indexing/pandas_normalized_tfidf.py:40]: What's the argument for removing this, shouldn't we keep this in?

Comment (chrispyl) [reply emm/indexing/pandas_normalized_tfidf.py:40]: Was thinking that it's connected to `PandasCosSimIndexer` which passes it's default `np.float32` `dtype` to it. Have put it back as default value in `kwargs` for use cases with different setups.

**Review by mbaak — APPROVED** [review]

File: emm/indexing/pandas_normalized_tfidf.py [hunk 2]
```diff
@@ -74,6 +73,8 @@ def fit(self, X: pd.Series | pd.DataFrame) -> TfidfVectorizer:
         with Timer("CustomizedTfidfVectorizer.fit") as timer:
             timer.label("super fit")
             super().fit(X)
+            # scikit-learn's TfidfVectorizer does not preserve dtype for large X, so we force it here
+            self.idf_ = self.idf_.astype(self.dtype)
```
Comment (glemaitre) [inline emm/indexing/pandas_normalized_tfidf.py:77]: So we have a fix in https://github.com/scikit-learn/scikit-learn/pull/30022.
Note that this only needed for NumPy < 2 then.

Comment (chrispyl) [reply emm/indexing/pandas_normalized_tfidf.py:77]: That's great, thank you for the quick investigation!

Comment (mbaak) [reply emm/indexing/pandas_normalized_tfidf.py:77]: Thanks @glemaitre, good to know!

**Pull request merged**